### PR TITLE
Domains and hosts API

### PR DESCRIPTION
**Pull request by kjac**
### Prerequisites

- [x] I have added steps to test this contribution in the description below

### Description

This PR adds domain and hosts editing to the management API.

### Testing this PR

Make sure you test:

- Setting default culture for a node
- Setting hosts for multiple cultures
- Setting the same host for different nodes (should not be possible)

@kjac has a Postman collection for testing.

### CODE REVIEW

**Review by Zeegaan — CHANGES_REQUESTED** [review]
Found a few minor things 👍

File: tests/Umbraco.Tests.Integration/Umbraco.Web.BackOffice/UrlAndDomains/DomainAndUrlsTests.cs [hunk 1]
```diff
@@ -124,6 +266,99 @@ public void Having_three_cultures_but_set_domain_on_a_non_default_language()
...
+        Assert.IsTrue(result.Success);
+
+        domains = domainService.GetAssignedDomains(Root.Id, true);
+        Assert.AreEqual(2, domains.Count());
+    }
+
+    [TestCase("/domain")]
+    [TestCase("/")]
+    [TestCase("some.domain.com")]
+    public async Task Cannot_assign_duplicate_domains(string domainName)
```
Comment (Zeegaan) [inline tests/Umbraco.Tests.Integration/Umbraco.Web.BackOffice/UrlAndDomains/DomainAndUrlsTests.cs:326]:
```suggestion
    public async Task Cannot_Assign_Duplicate_Domains(string domainName)
```

File: tests/Umbraco.Tests.Integration/Umbraco.Web.BackOffice/UrlAndDomains/DomainAndUrlsTests.cs [hunk 2]
```diff
@@ -124,6 +266,99 @@ public void Having_three_cultures_but_set_domain_on_a_non_default_language()
...
+            Domains = Cultures.Select(culture => new DomainModel { DomainName = domainName, IsoCode = culture }).ToArray()
+        };
+
+        var result = await domainService.UpdateDomainsAsync(Root.Key, updateModel);
+        Assert.IsFalse(result.Success);
+        Assert.AreEqual(DomainOperationStatus.DuplicateDomainName, result.Status);
+    }
+
+    [Test]
+    public async Task Cannot_assign_already_used_domains()
```
Comment (Zeegaan) [inline tests/Umbraco.Tests.Integration/Umbraco.Web.BackOffice/UrlAndDomains/DomainAndUrlsTests.cs:340]:
```suggestion
    public async Task Cannot_Assign_Already_Used_Domains()
```

File: tests/Umbraco.Tests.Integration/Umbraco.Web.BackOffice/UrlAndDomains/DomainAndUrlsTests.cs [hunk 3]
```diff
@@ -74,17 +76,147 @@ protected override void CustomTestSetup(IUmbracoBuilder builder)
     private readonly TestVariationContextAccessor _variationContextAccessor = new();
 
     public IContent Root { get; set; }
+
     public string[] Cultures { get; set; }
 
+    [Test]
+    public async Task Can_update_domains_for_all_cultures()
```
[the remark this answers is not in the document]
Comment (kjac) [reply tests/Umbraco.Tests.Integration/Umbraco.Web.BackOffice/UrlAndDomains/DomainAndUrlsTests.cs:83]: Yeah I just went with how it was before I changed things 😄

File: src/Umbraco.Core/Services/DomainService.cs [hunk 4]
```diff
@@ -1,23 +1,55 @@
...
+using Umbraco.Cms.Core.Services.OperationStatus;
+using Umbraco.Extensions;
 
 namespace Umbraco.Cms.Core.Services;
 
 public class DomainService : RepositoryService, IDomainService
 {
     private readonly IDomainRepository _domainRepository;
+    private readonly ILanguageService _languageService;
+    // private readonly IEntityService _entityService;
```
[the remark this answers is not in the document]
Comment (kjac) [reply src/Umbraco.Core/Services/DomainService.cs]: Fixed 😄

File: src/Umbraco.Core/Services/DomainService.cs [hunk 5]
```diff
@@ -27,26 +59,15 @@ public bool Exists(string domainName)
...
-            _domainRepository.Delete(domain);
-            scope.Complete();
-
-            scope.Notifications.Publish(new DomainDeletedNotification(domain, eventMessages).WithStateFrom(deletingNotification));
-        }
-
-        return OperationResult.Attempt.Succeed(eventMessages);
+        return DeleteAll(new[] { domain }, scope, eventMessages)
+            ? OperationResult.Attempt.Succeed(eventMessages)
+            : OperationResult.Attempt.Cancel(eventMessages);
```
[the remark this answers is not in the document]
Comment (kjac) [reply src/Umbraco.Core/Services/DomainService.cs]: Fixed

File: src/Umbraco.Core/Services/DomainService.cs [hunk 6]
```diff
@@ -81,28 +104,19 @@ public IEnumerable<IDomain> GetAssignedDomains(int contentId, bool includeWildca
...
-                scope.Complete();
-                return OperationResult.Attempt.Cancel(eventMessages);
-            }
-
-            _domainRepository.Save(domainEntity);
-            scope.Complete();
-
-            scope.Notifications.Publish(new DomainSavedNotification(domainEntity, eventMessages).WithStateFrom(savingNotification));
-        }
+        using ICoreScope scope = ScopeProvider.CreateCoreScope(autoComplete: true);
```
[the remark this answers is not in the document]
Comment (kjac) [reply src/Umbraco.Core/Services/DomainService.cs]: Fixed!

File: tests/Umbraco.Tests.Integration/Umbraco.Web.BackOffice/UrlAndDomains/DomainAndUrlsTests.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (kjac) [reply tests/Umbraco.Tests.Integration/Umbraco.Web.BackOffice/UrlAndDomains/DomainAndUrlsTests.cs]: Fixed all of them 😄

**Comment by Zeegaan** [comment]
Looks good, tests good 🚀

**Review by Zeegaan — APPROVED** [review]

**Pull request merged**